### PR TITLE
Implement SPV_KHR_ray_query generation

### PR DESCRIPTION
**Pull request by jiaolu**
from dxr 1.1 rayquery hlsl

### CODE REVIEW

**Comment by ehsannas** [comment]
@jiaolu Please check the test failures.

**Comment by jiaolu** [comment]
> @jiaolu Please check the test failures.
Is there way to run single failed test? .e.g . my one failed test : WholeFileTest.PassThruPixelShader ?

**Comment by ehsannas** [comment]
run 
```
clang-spirv-tests.exe --spirv-test-root C:\DirectXShaderCompiler\tools\clang\test\CodeGenSPIRV --gtest_filter= WholeFileTest.PassThruPixelShader
```

**Comment by jiaolu** [comment]
> Please check the test failures.
@ehsannas, The checks have passed.

**Comment by jiaolu** [comment]
@ehsannas, Any comment on this review?

**Comment by ehsannas** [comment]
Sorry, I'm falling behind in code reviews. I'll definitely review this as soon as I can.

File: docs/SPIR-V.rst [hunk 1]
```diff
@@ -3282,6 +3282,140 @@ Interface Variables
...
+Shaders can instantiate RayQuery objects as local variables, the RayQuery object acts as a state
+machine for ray query. The shader interacts with the RayQuery object’s methods to advance the
+query through an acceleration structure and query traversal information
+
+| Refer to following pages for details:
+| https://microsoft.github.io/DirectX-Specs/d3d/Raytracing.html
+
+A flow chart for a simple ray query process
+
+          +------------------------------+
```
Comment (ehsannas) [inline docs/SPIR-V.rst:3299]: I think you need to add `::` followed by an empty line before this line

Comment (ehsannas) [reply docs/SPIR-V.rst:3299]: See line 2928 as an example

File: docs/SPIR-V.rst [hunk 2]
```diff
@@ -3282,6 +3282,140 @@ Interface Variables
...
+                       |
+                       V
+        +----------------------------------+
+        | Your Intersection/shader code    |
+        +----------------------------------+
+
+
+.. Example flow chart code:: hlsl
+
+void main() {
```
Comment (ehsannas) [inline docs/SPIR-V.rst:3333]: need to have `.. code:: hlsl` followed by an empty line before this line.
See line 2975 as an example.

**Comment by dgkoch** [comment]
> Do you recommend waiting for the extension to become final first?

We'd like to have this available before final so that people can experiment with it.
Obviously it'll need to be updated to match final when that is available.

**Comment by ehsannas** [comment]
> We'd like to have this available before final so that people can experiment with it.
Obviously it'll need to be updated to match final when that is available.

@dgkoch Sounds good to me.

@jiaolu Please add the following warning to the beginning of the two functions `processTraceRayInline` and `processRayQueryIntrinsics`:

```cpp
emitWarning("SPV_KHR_ray_query is currently a provisional extension and might
 change in ways that are not backwards compatible", expr->getExprLoc());
```

**Comment by jiaolu** [comment]
code updated.

**Comment by jiaolu** [comment]
yes, i have tested dxc output of rayquery spv in our internal setup environment.

**Comment by jiaolu** [comment]
@ehsannas,  Any more comments about this review?

**Comment by ehsannas** [comment]
Thanks @jiaolu . I have reverted all the changes to HlslTypes.cpp because it was just spacing changes and it was distracting from the main goal of this PR (sadly some files within DXC currently don't adhere to clang-format). The formatting of the doc file was also still incorrect which resulted in Github not being able to render the file correctly. I've fixed those too.

**Comment by jiaolu** [comment]
@ehsannas , Thanks for the fix

**Pull request merged**